plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -296,53 +316,43 @@ public Map<String, BlobMetadata> listBlobs() throws IOException {
-            return executeListing(clientReference, listObjectsRequest(keyPath)).stream().flatMap(listing -> {
-                assert listing.getObjectSummaries().stream().noneMatch(s -> {
-                    for (String commonPrefix : listing.getCommonPrefixes()) {
-                        if (s.getKey().substring(keyPath.length()).startsWith(commonPrefix)) {
+            return executeListing(clientReference, listObjectsRequest(keyPath)).stream().flatMap(listObjectsResponse -> {
+                assert listObjectsResponse.contents().stream().noneMatch(s -> {
+                    for (CommonPrefix commonPrefix : listObjectsResponse.commonPrefixes()) {
+                        if (s.key().substring(keyPath.length()).startsWith(commonPrefix.prefix())) {
                             return true;
                         }
                     }
                     return false;
                 }) : "Response contained children for listed common prefixes.";
-                return listing.getCommonPrefixes().stream();
+                return listObjectsResponse.commonPrefixes().stream();
             })
-                .map(prefix -> prefix.substring(keyPath.length()))
+                .map(commonPrefix -> commonPrefix.prefix().substring(keyPath.length()))
                 .filter(name -> name.isEmpty() == false)
                 // Stripping the trailing slash off of the common prefix
                 .map(name -> name.substring(0, name.length() - 1))
                 .collect(Collectors.toMap(Function.identity(), name -> blobStore.blobContainer(path().add(name))));
-        } catch (final AmazonClientException e) {
+        } catch (final RuntimeException e) {
             throw new IOException("Exception when listing children of [" + path().buildAsString() + ']', e);
         }
     }
 
-    private static List<ObjectListing> executeListing(AmazonS3Reference clientReference, ListObjectsRequest listObjectsRequest) {
-        final List<ObjectListing> results = new ArrayList<>();
-        ObjectListing prevListing = null;
-        while (true) {
-            ObjectListing list;
-            if (prevListing != null) {
-                final ObjectListing finalPrevListing = prevListing;
-                list = SocketAccess.doPrivileged(() -> clientReference.get().listNextBatchOfObjects(finalPrevListing));
-            } else {
-                list = SocketAccess.doPrivileged(() -> clientReference.get().listObjects(listObjectsRequest));
-            }
-            results.add(list);
-            if (list.isTruncated()) {
-                prevListing = list;
-            } else {
-                break;
-            }
-        }
+    private static List<ListObjectsV2Response> executeListing(AmazonS3Reference clientReference, ListObjectsV2Request listObjectsRequest) {
+        final List<ListObjectsV2Response> results = new ArrayList<>();
+        ListObjectsV2Iterable listObjectsIterable = SocketAccess.doPrivileged(
```

Comment:
Shouldn't we combine these to calls into one privileged block?

Reply:
We can do this yes, I'll push an update.